File: problems/76/76.minimum-window-substring.go
/*
 * @lc app=leetcode id=76 lang=golang
 *
 * [76] Minimum Window Substring
 */

package solution

import "math"

// @lc code=start
func minWindow(s string, t string) string {
	m := len(s)
	n := len(t)

	if m == 0 || n == 0 || m < n {
		return ""
	}

	dict := make(map[byte]int)
	for i := 0; i < n; i++ {
		dict[t[i]]++
	}

	required := len(dict)
	actual := 0
	window := make(map[byte]int)
	res := math.MaxInt64
	res_l, res_r := 0, 0
	l := 0

	for r := 0; r < m; r++ {
		curr_char := s[r]
		window[curr_char]++

		if cnt, ok := dict[curr_char]; ok && window[curr_char] == cnt {
			actual++
		}

		for l <= r && actual == required {
			curr_size := r - l + 1
			if curr_size < res {
				res = curr_size
				res_l, res_r = l, r
			}

			left_char := s[l]
			window[left_char]--

			if cnt, ok := dict[left_char]; ok && window[left_char] < cnt {
				actual--
			}

			l++
		}
	}

	if res == math.MaxInt64 {
		return ""
	}

	return s[res_l : res_r+1]
}

// @lc code=end

File: problems/76/76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
from collections import Counter, defaultdict


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # return the minimum window substring of s such that
        # every character in t (including duplicates) is included in the window
        # ie:
        # s = "ABBBBABC", t = "ABC"

        if not t or not s:
            return ""

        required_chars = Counter(t)
        curr_chars = 0
        curr_window = defaultdict(int)

        res = (float("inf"), -1, -1)
        l, r = 0, 0

        while r < len(s):
            curr_char = s[r]
            curr_window[curr_char] += 1

            if (
                curr_char in required_chars
                and curr_window[curr_char] == required_chars[curr_char]
            ):
                curr_chars += 1

            while l <= r and curr_chars == len(required_chars):
                left_char = s[l]

                if r - l + 1 < res[0]:
                    res = (r - l + 1, l, r)

                curr_window[left_char] -= 1
                if (
                    left_char in required_chars
                    and curr_window[left_char] < required_chars[left_char]
                ):
                    curr_chars -= 1

                l += 1

            r += 1

        return "" if res[0] == float("inf") else s[res[1] : res[2] + 1]


# @lc code=end
